### ISSUE TITLE
Reducing block sizes to allow for more detailed structures.

### ISSUE THREAD

**Issue by ghost** [minetest/minetest#3376]
A while back I worked on a proposal about how it might be possible to reduce the block sizes in minetest.  I was hoping to get the interest of some irrlicht programmmers.
However, I have kind of abandoned the project...  Still I figured maybe it would be best to just post what I had so far instead of letting it get lost forever.  Maybe someone out there would be interested in it.

Sidenote: Don't bother asking me questions about it, as I won't be able to respond in the future.  However, if the minetest developers or a few competent programmers want to discuss and consider it, great!

**Theory**
Think for a moment about another toy: Legos.  With them, you have many small blocks that you can build almost an infinite number of things.  For this reason, they can be played with for many years.  Now compare them to Duplos ... large blocks made for small kids.
I like to compare minetest to Duplos; it has the same type of oversized blocks.  However, they suffer from the same problem as Duplos... eventually you run out of things you can make/build with them and they become very boring very fast.  (Although I think we will all admit that minetest can be enjoyed for much longer.)
What I am proposing is that we "need" to make the block sizes smaller to make minetest have a much longer play value... and to allow minetest to be used to make many things that are simply impossible with the current large block sizes.  I do think this could bring in a number of more serious builders to the game as well ... and also result in some really great things being built that are not currently possible.
The following is a proposal for block sizes that are half the current size (thus 8x more blocks).  Smaller blocks sizes would be ideal, but I think maybe only half the current size is possible without major rewrites to some parts of minetest.

**Networking**
Cost
1. 8x as much data transmitted to players
2. up to 8x as much data transmitted to server when a player builds or digs
Solution
I don't see a way to completely solve this, but there are a number of ways that might reduce networking traffic.
1. Use a compression algorithm specifically designed for "real-time" compression on servers to send compressed map data to players.
2. Group block changes sent to the server (especially when a player digs or places an entire 1m cube... transmit one request instead of 8).
Details:
1. Instead of sending raw map chuncks to the players, compress the data on the fly using one of the many "real-time" compression libraries designed for high speed compression.
Examples:
snappy: https://code.google.com/p/snappy/
LZ4: https://github.com/Cyan4973/lz4
Perhaps something like this:
1.	Group 16x16x16 chuncks together into groups of maybe 8 or some other number... (so you have 32x32x32).  The goal here is that you want enough data that compression actually helps.
2.	Continue working with data on the server machine uncompressed in RAM.
3.	Before sending the group to a player, compress it in real-time using LZ4.
4.	The player receives it and decompresses the data in real-time.
2. As for the player sending data to the server, there is one simple technique that could help a little bit.  Since we have subdivided a 1m cube into 8 smaller cubes... anytime a player digs or places an entire 1m cube, you could send one request instead of 8 separate ones.  Of course, this does not solve many other issues, but it would keep networking usage closer to current levels under certain conditions.
**Mapgen**
Cost
The mapgen would be 8x slower!
Solution
The solution is rather easy: make the mapgen multithreaded.
Details:
I have studied mapgens in great detail over the months and am almost certain that I could make a mapgen that is:
⦁	100% determinstic (or close enough that it won't matter much)
⦁	utilizes all cores of a CPU in parallel
Since I know for certain this can be done (I know how to write such a mapgen arleady), I will not go into the details.  Using something like lua lanes for the parallel threads would be my preference: http://lualanes.github.io/lanes/ (if the mapgen is LUA, that is)
**Storage**
Cost
8x as much hard drive space...  good servers can use 4GB to 10GB ... so that's 32 to 80GB!
Solution
Reducing the amount of hard drive space involves a number of changes.  However, it will basically require a NEW storage format.
1. Remove param1 data (lighting) from the database.  Replace with minimal lighting info in the database and on-demand calculation of lighting by each client (or the server).
2. Do not place node name strings inside each chunck of the database.  Instead use a global lookup table for the entire database, with the name strings defined only one time in the entire database.
3. Group map chuncks into large enough groups that it can be compressed effectively... and then either compress the data and store inside the database or use LevelDB, which automatically compresses with snappy compression.
4. Use Huffman encoding (before compression is used) or some other technique to store "shape" data inside the database.
Details:
1. From what I can tell, lighting calculations only take a few milliseconds for a portion of the map (For example, the voxel manip lighting calcuations are very fast).  Thus, could we just take the current lighting algorithm as-is, with minimal changes, and make each client calculate lighting whenever they receive a chunck of map from the server?  Each client could then store lighting data internally (in RAM) once it's calculated, of course.  As far, as I know this would be possible to do except for a few problems:
⦁	It is EXTREMELY important that all clients end up calucating the same lighting (you don't want some players seeing dark and the others seeing daytime) because different clients have different chuncks downloaded from the server.
⦁	One key problem is that lighting relies on information stored in the map to know about things like.... if a mining shaft is connect to the surface many thousands of blocks up... or whether there is a building high up... and whether it should show daylight.
Thus, some changes will be needed.  Here's a few possibilities:
⦁	It may be necessary to simplify or change the lighting to be less accurate.  For example, that mineshaft will no longer be able to propogate lighting so far down.
⦁	Instead of storing lighting data (param1) inside the database for all blocks, only store lighting info for the edges of groups of blocks.  For example, let's say we group blocks into 32x32x32 ... you might only store lighting for the bottom layer (32x32) inside the database so then the game knows if light is propagating downwards from chuncks above.  All the rest could be calculated client-side.  (Sidenote: it may be necessary to include lighting information on each of the sides too, but if possible it would be ideal to avoid that.)
2. Currently, minetest stores the entire string name for nodes inside each chunck.  For huge maps, this is a lot of repeated data taking up space.  One simple solution would be to create an extra entry inside the database that only stores the string names ... and their corresponding shorter code.  Then each map chunck uses the shorter code.
3. An obvious solution is to compress the data being stored into the database.  From what I can tell levelDB does this automatically using snappy compression.  However, it may be necessary to group the map database into larger chuncks that can be compresses more effectively.  Another alternative is to use LZ4 compression on a simple key/value store.
4. I would like to see "shape" data added to the game. (Instead of having to create totally new blocks, include the ability to make any block into various shapes, triangles, half squares, and many others).  The simplest solution is to use a fixed number of bits to store shape data (like 8 bits for lighting), but this is too wasteful.
⦁	Option 1: Use Huffman encoding for shape data.  This allows the shapes to use up only the amount needed and no more... It also allows you to add an infinite number of new shapes to the game via mods without changing the map format.  It is a bit wasteful on space though.
⦁	Option 2: Used a fixed number of bits inside the database, but limit the number of bits uses to the maximum amount of shapes in that particular map/game  The good news is that mods could still add an unlimited number of extra shapes to the game without a core change or map format change.  The bad news is that the entire map database would have to be "processed" to add the extra bits for that particular game/map.
Note: This shape format occurs before the compression process.

I never finished it........

Another important aspect I was going to include in the proposal but never got around to writing:
With smaller blocks sizes, you can't use a single repeating texture for each block... the reaping patter would look aweful.
You would need a way to map a larger texture onto multiple blocks. You could do something like having a 16x16 textures that maps onto 2x2 1/2 meter squares ... basically preserving the current texture layout. Additionally, it could be expanded further to map larger size textures (like 64x64) onto an arbritrary number of cubes, like making a 64x64 texture map onto 8x8 1/2 meter squares.

BTW, one possible objection might be that smaller blocks would require better client computers (some minetest players don't have such computers).  This is quite true... however, a lot of people with better computers have already left games like minetest due to poor graphics and blocks sizes that do not give enough control over building... so, in fact, smaller block sizes would attract new players... ones who have computers that can handle the game.  Choosing to, instead, keep large blocks sizes only because it allows slower computers to handle game actually reinforces and keep minetest limited to players with slower computers.

**Comment by paramat** [comment]
This would have to be a fork of Minetest, it's too much of a fundamental change to considered.

**Comment by est31** [comment]
First, the map format has no chunks. We have mapblocks for what minecraft calls chunks. They are different concepts though, a mapblock doesn't extend from the world's top border to the world's bottom border.

Your comment touches alot of issues.
Some of them are valid ideas, which we already considered, but some of them are better not done, or have a better place in a fork.
To answer to a few points:
* duplo vs lego: We could add support for making the block size smaller, e.g. by enabling to scale the player entity. This would be quite fun. However, I don't see why it would need that many of the changes you list, like sending more stuff or so. Scaling the player entity is enough, that involves a single float to be synched between client and server, nothing more.
* threaded mapgen: The c++ mapgen is already threaded AFAIK except for locks, and hmmm is working on making lua mapgen threaded as well. hmmmm also already thinks of solutions to make the lock granularity smaller
* deterministic mapgen: on my wishlist for a very long time, but I guess nobody wants to be the mapgen guy once we have it.
* don't store lighting for every node: The idea with storing on the borders has already been mentioned in discussions, multiple times. It doesn't solve some of the problems we have with the current lighting model, like that it doesn't cope with fast updates (voxel SAO moving etc)
* shape idea: Doesn't bring any usefulness. You can achieve it already with current means, you can use as many bits of param0 as you want. Just register a lua loop that calls minetest.register_node repeated times. This is far more flexible than to hard-code the list of shapes. Huffman coding shapes doesnt help at all, you still would need to agree on a tree.
* db-global nameidmapping: This isn't an idea that hasn't been considered already as well. I had done rough calculations: db-global nameidmapping would indeed spare space, however that approach would work worse when the set of mods changes. Before, you only had unknown nodes; now you have to cope with a potentially totally different nameidmapping. Also, the format would be slightly more complex, I would however agree to a change if it solves all the problems, some of which mentioned above.
* compression of mapblock data: its already done. we compress with zlib.
* grouping more stuff together than just a mapblock in order to have even more efficient compression: I had this idea as well, but didn't really follow it, to do actual testing with real world data. It has been my conclusion of my experiments for using lossles mode of vp9 to compress param1 data.

I'll close this issue. If somebody in the dev team thinks that there is any value in having this issue open, feel free to reopen.

**Comment by 0-afflatus** [comment]
Cube.

**Comment by C1ffisme** [comment]
I actually want this to happen, although I know this would take years of work to not break every minetest server and world ever.

**Comment by asl97** [comment]
@est31 
>db-global nameidmapping: This isn't an idea that hasn't been considered already as well. I had done rough calculations: db-global nameidmapping would indeed spare space, however that approach would work worse when the set of mods changes. Before, you only had unknown nodes; now you have to cope with a potentially totally different nameidmapping. Also, the format would be slightly more complex, I would however agree to a change if it solves all the problems, some of which mentioned above.

why would the nameidmapping be totally different?

the nameidmapping should be store in the world database since it would be unique to that world. the logic should be something like:

* on register node should check if node is in the nameidmapping, if it is in it, it get the id from it, otherwise it add it in with a new id

* if an id can't be map to a node (which should never happen) or is mapped to an unregistered node, it would be consider as an unknown node.

making it backward compatible might be complex and messy but without backward compatible, it shouldn't be that much more complex.
____
i have to question how much space it would save though.  currently the block size seem to be related to the number of different node in it.

having a global nameidmapping that use a fix amount of bit would mean the amount of space use in most of not all cases would increase and it would also add a limit on the amount of nodes

i doubt a global nameidmapping which doesn't use a fix amount of bit would work great.
